File: 0141-linked-list-cycle/141.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # Approach: Two Pointer -- T.C: O(N), S.C: O(1)
        slow, fast = head, head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
        return False

        """
        # Approach: HashSet -- T.C: O(N), S.C: O(N)
        curr = head
        visited_addr = set()

        while curr:
            curr_addr = id(curr)
            if curr_addr in visited_addr:
                return True
            visited_addr.add(curr_addr)
            curr = curr.next
        return False
        """

File: 0141-linked-list-cycle/141.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
  public:
    bool hasCycle(ListNode *head) {
        /* Approach:
            1. Fast and Slow Pointer -- T.C: O(N), S.C: O(1)
        */

        if (head == nullptr)
            return false;

        ListNode *fast = head, *slow = head;

        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast)
                return true;
        }

        return false;
    }
};
